src/core/Oscillator.cpp

```diff
@@ -474,37 +724,65 @@ template<>
 inline sample_t Oscillator::getSample<Oscillator::SineWave>(
 							const float _sample )
 {
-	return( sinSample( _sample ) );
+	if (m_useWaveTable)
```

Comment:
It's not alias free if the frequency is higher than the Nyquist frequency.

Reply:
OK, but it still should not require pre-computation. An `if(freq <= nyquist) { play(); }` should suffice.

Reply:
Makes sense.

Reply:
There is one problem with the `if (...) play()` approach: all the functions that operate on wave shapes are universal templates that work the same way for all wave shapes. So `SineWave` would probably need a specialization for all of these? (Or maybe not, I'm not sure yet..) If that is the case, it would just make the code more complicated for not much benefit.

Reply:
Also, this "not playing above Nyquist" (or possibly just "not playing above 20 kHz) is already working in TripleOsc, so it would be sufficient to simply remove the sine wavetable and always return the standard generated sample.

The problem is that other instruments that will possibly use these oscillators in the future may not enforce such limit. So the check would still need to be added to ... well, somewhere. Or everywhere? Since m_freq is used all over the place and there is no single obvious "play()" function. There is `update()`, which already seems to contain this check, but it apparently does not prevent Monstro or LB302 from playing sine waves above Nyquist. So either `update()` does something different than I think, or the instruments use the oscillator in some custom way that bypasses the check.

Either way I don't really know where would I add another check that achieves what we want (to avoid using sine wavetable, while preventing playing sines above Nyquist).

Reply:
Not sure which specialization you plan to add - you are already inside of one. I would just have removed the `if` branch (I think that's what you propose in the 1st paragraph of your 2nd reply), so that it always just returns `sinSample`.

I would have thought that a simple check against the nyquist freq would be to check that `m_freq * m_detuning * Engine::mixer()->processingSampleRate()` (that's the term passed to `waveTableBandFromFreq`) is `<= OscillatorConstants::MAX_FREQ`.

Reply:
Yes, `getSample()` for sine wave would always return `sinSample()`. And there should be a check that prevents it from returning anything if the sample is requested for frequency higher than ~20 kHz. That is clear. What I'm not sure about is _where_ should the check go?

There is already one in `Oscillator::update()`, which is probably why TripleOsc does not play any notes above Nyquist, as defined by the current sample rate. So all that would be needed is changing it so that it checks against `OscillatorConstants::MAX_FREQ` when wavetable mode is enabled.

But with all the sub-oscillator business, phase modulation and other shenanigans, I am not sure if that would be sufficient. I.e., could it happen that some sub-oscillator generates signal at 30 kHz and messes things up by reflecting down to around 10 kHz? In the wavetable mode, it simply could not generate any signal above 20 kHz, whereas `sinSample()` always happily returns a value.

I _think_ it should be fine and the check in `Oscillator::update()` would be enough, but remember that I just fixed a few specific issues in the original PR, I don't really know how all the things in `Oscillator` work and how they interact with the instruments. For example, the check in `Oscillator::update()` prevents TripleOsc from playing sine waves above Nyquist, but it does not prevent the same from happening in Monstro and LB302 (if I tested it correctly), and I have no idea why. So I can't say whether I'm missing something that would make the sine wave shape ignore the band limits in some situations if I just blindly return `sinSample()` and check the frequency limit only in `update()`.

Reply:
I thought to do it right here, no in update. Currently, your code calls `wtSample`, which calls `waveTableBandFromFreq`, which is equivalent to a check against nyquist freq, right? (because `waveTableBandFromFreq` picks that band that we need, and it picks a zero wave if the freq is too high). So instead, if you just add a check exactly here, shouldn't it trigger exactly if you current check triggers?

Hopefully I'm not overcomplicating. The idea was just to save 1 or 2 MB of cache for the sine waves, but if this discussion gets too complex, feel free to resolve :smiley:

Reply:
> I thought to do it right here, no in update. Currently, your code calls wtSample, which calls waveTableBandFromFreq, which is equivalent to a check against nyquist freq, right?

Oh, damn, I think you are right. It is this weird design where `m_freq` is _the_ frequency that is currently playing. I did not even realize it means that it is always available to check from any place, even from `wtSample()`. Thanks for patiently pushing my brain to finally get it. :smile:  I will test it out, but it seems to me it should work.

Reply:
Hm, so removing the table for `SineWave` is not so straight-forward. I moved it to the end of the `WaveShapes` enum so that the shapes that _do_ use pre-generated band-limited shapes start at zero, but now TripleOsc plays triangle wave when it should play sine.

Which tells me, that the TripleOsc probably ignores the enum and just blindly assigns integer from the GUI switch, assuming index zero is sine. I guess the best way to solve this will be to fix TripleOsc (and check if other instruments do not make a similar mistake).

Reply:
Uhm, never mind. It seems this kind of change would require an update routine, because the index in project files and presets would not match either. Suddenly it does not seem as the best solution. So maybe I'll just make the table of wave shapes one element smaller and make sure `Oscillator` always subtracts one from the index. Since the table is not used anywhere else, it should be OK-ish.